tests: Port thread_metric benchmark from ThreadX

peter-mitsis opened the pull request:
Ports the thread_metric suite of benchmarks from ThreadX to Zephyr. This port was prompted by the recent RTOS performance comparison paper by Jacob Beningo which showed Zephyr as having poor performance relative to other RTOSes.

Having our own port of this benchmark suite should make it easier for others to use and reference AND maintain an optimal prj.conf.

Note: This code uses an MIT license.

peter-mitsis committed:
tests: Port Thread-Metric benchmark from ThreadX

Ports the Thread-Metric suite of benchmarks from ThreadX to Zephyr.
This makes it easier for others to run these benchmarks with the
best set of configuration options for their board so that
they can better compare Zephyr performance to another RTOS.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

nashif commented:
would be nice to maintain the flat tree structure as with the origin and do all the magic in one single cmake file, similar to what we do in philosopher sample, i.e. using a compile definition to select the relevant sources.

This will allow us to easily update in case the origin has change.

peter-mitsis commented:
Flattened the directory structure.

nashif commented:
needs exception due to license

henrikbrixandersen reviewed (changes requested):
Please add an entry in https://docs.zephyrproject.org/latest/LICENSING.html#zephyr-licensing for this.

nashif commented:
> Please add an entry in [docs.zephyrproject.org/latest/LICENSING.html#zephyr-licensing](https://docs.zephyrproject.org/latest/LICENSING.html#zephyr-licensing) for this.

@peter-mitsis  can you please add that.

peter-mitsis committed:
doc: Update licensing page

Updates the licensing page to indicate that the thread_metric
benchmark uses the MIT license.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

kartben commented on tests/benchmarks/thread_metric/src/tm_porting_layer_zephyr.c:
```diff
@@ -0,0 +1,222 @@
+/***************************************************************************
+ * Copyright (c) 2024 Microsoft Corporation
```
Missing Intel copyright

peter-mitsis replied:
Done.

kartben reviewed (changes requested):
Some suggested changes to use naming consistent with the upstream project (including Eclipse ThreadX attribution in the name of the main Kconfigs as I think it's fair game), plus a missing copyright entry that needs fixing 
Thanks a lot for this work, @peter-mitsis!

kartben commented on tests/benchmarks/thread_metric/Kconfig:
```diff
@@ -0,0 +1,83 @@
+	prompt "Select the thread metric to execute"
+	default TM_PREEMPTIVE
+	help
+	  The thread metric benchmark suite consists of eight RTOS tests.
```
If there are other instances of "thread metric" I missed please consider fixing too

peter-mitsis replied:
Done.

peter-mitsis commented on tests/benchmarks/thread_metric/Kconfig:
```diff
@@ -0,0 +1,83 @@
+# Copyright (c) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+mainmenu "Latency Measure Benchmark"
```
Done.

peter-mitsis commented on tests/benchmarks/thread_metric/Kconfig:
```diff
@@ -0,0 +1,83 @@
+source "Kconfig.zephyr"
+
+choice TM_TEST
+	prompt "Select the thread metric to execute"
```
Done.

nashif commented:
@peter-mitsis  can you please run a full twister on this, i.e.

``` ./scripts/twister -T tests/benchmarks/thread_metric/ -v```

I see a few failures on some platforms, this might be sporadic behavious and will be a problem when merged.

peter-mitsis commented:
I ran twister on my machine and observed two timeout errors. The first was on qemu_malta/qemu_malta and the second on qemu_malta/qemu_malta/be.  In both instances it was the CONFIG_TM_MESSAGE=y testcase and both their handler.log files only contained the "*** Booting Zephyr OS ..." line. When I ran just those configurations, they passed. I think that the timeout errors were related to system load.  These tests are supposed to take about 30 seconds before displaying the output messages.

peter-mitsis commented:
Timeout for the tests were increased to account for the timeouts due to host load and the approximate 30 second (best case) delay before there is any output.
